MAINT: backports for SciPy 1.17.0rc2

tylerjereddy opened the pull request:
Note that this set of backports is quite large, and even includes two enhancement PRs to accommodate deprecations/plans in the array API space, so it may increase the chances of needing an RC3 (which I'm prepared to deal with, though still have my fingers crossed). 

* Update the SciPy `1.17.0` release notes following additional backport activity, including updates to the author list and issue/PR lists.  
* Address the comment at: https://github.com/scipy/scipy/pull/24138#discussion_r2615964668  
* Fix a few typos in the release notes that were noticed after RC1.
* Address the missing release notes entries based on the summary at: https://github.com/scipy/scipy/issues/24152

Backports included here (so far):

1. gh-24030
2. gh-24031
3. gh-24132
4. gh-24133
5. gh-24139
6. gh-24140
7. gh-24143
8. gh-24145
9. gh-24151
10. gh-24155
11. gh-24161
12. gh-24174
13. gh-24213
14. gh-24223
15. gh-24218
16. gh-24111 (pulled in the `pixi.toml` changes, discarded lock file changes, then ran `pixi lock`)
17. gh-24239
18. gh-24240

TODO:

- [x] ensure the regular CI is passing
- [x] probably do a dry-run of the wheel builds, I usually do that before starting the rel process proper
- [x] is there anything else left to address with `1.17.0` milestone (I think we're "ok" for those items after this backport PR?): https://github.com/scipy/scipy/issues?q=is%3Aissue%20state%3Aopen%20milestone%3A1.17.0

JaRoSchm committed:
DOC: update mailmap

[skip ci]

j-bowhay committed:
BUG: integrate: fix no module named 'scipy.integrate._lsoda' (#24133)

mdhaber committed:
ENH: stats.anderson: add `method` parameter (#24030)

* ENH: stats.anderson: add method parameter

* TST: stats.anderson: test method parameter

* MAINT: stats.anderson: adjustments per self-review

* DOC: stats.anderson: adjust documentation

jorenham committed:
API: ``signal.savgol_coeffs``: keyword-only ``xp`` and ``device`` (#24140)

jorenham committed:
API: ``linalg.inv``: keyword-only ``assume_a`` and ``lower`` (#24139)

jorenham committed:
TYP: generic ``signal.LinearTimeInvariant`` type (#24143)

mdhaber committed:
ENH: stats.anderson_ksamp: add 'variant' parameter (#24031)

jorenham committed:
BUG: ``sparse``: fix ``coo_matrix.__setitem__`` signature (#24145)

mdhaber committed:
MAINT: linalg: raise for zero-size batch

mdhaber committed:
STY: _lib._util: use type instead of TypeAlias

[lint only]

fbourgey committed:
MAINT: bump `xsf` to fix `special.itj0y0` regression (#24155)

ilayn committed:
BUG:linalg.interpolative:Care for single row arrays

ilayn committed:
BUG:linalg.interpolative:Prepare nonarray args before sending to idd_reconid

ilayn committed:
REV:integrate: Revert the stepsize change in LSODA

tylerjereddy committed:
DOC: update the 1.17.0 release notes

* Update the SciPy `1.17.0` release notes
following additional backport activity, including
updates to the author list and issue/PR lists.

* Address the comment at:
https://github.com/scipy/scipy/pull/24138#discussion_r2615964668

* Fix a few typos in the release notes that were noticed
after RC1.

* Address the missing release notes entries based on the summary
at:
https://github.com/scipy/scipy/issues/24152

tylerjereddy committed:
MAINT, TYP: PR 24207 revisions [wheel build]

* manually revert a few typing-related adjustments
based on reviewer feedback at:
https://github.com/scipy/scipy/pull/24207/changes#r2637006039

[wheel build]

j-bowhay reviewed (dismissed):
Can you add the two new deprecations for stats to the release note too

steppi committed:
TST: fft: More allow_dask_compute=True (#24213)

tylerjereddy committed:
BENCH, MAINT: NumPy 2.4.0 bench compat (#24218)

* Fixes gh-24212.

* The changes here allow the problematic benchmarking incantation described
in the matching ticket to have a passing/zero exit code with NumPy
`2.4.0`:
`asv run --quick -b "time_fancy_getitem" -e`

* I looked at the circleci failure log in the matching issue and this
seems to be the only issue with NumPy `2.4.0` in the benchmark suite,
but we'll see what the CI says..

lucascolley committed:
BLD: linalg: `link_language: 'fortran'` for `_fblas` (#24223)

lucascolley committed:
BLD/DEV: use flang 21 on Windows (#24111)

[skip circle]

tylerjereddy committed:
DOC: PR 24207 revisions

* Update the SciPy `1.17.0` release notes following additional
backport activity. Also mention the two `stats` deprecations
requested (https://github.com/scipy/scipy/pull/24207#pullrequestreview-3601342641).

* Also address release notes review comments from:
https://github.com/scipy/scipy/pull/24207#discussion_r2637390315
https://github.com/scipy/scipy/pull/24207#discussion_r2637390386

lucascolley committed:
MAINT: bump array-api-compat for JAX>=0.8.2 compat

lucascolley committed:
MAINT: bump array-api-compat to 1.13 tag

tylerjereddy committed:
CI: PR 24207 revisions

* Pin `scipy-openblas` versions in a problematic CI
job to avoid issues with a recent release.

tylerjereddy committed:
DOC: PR 24207 revisions

* Update the SciPy `1.17.0` release notes
following additional backport activity.

jorenham reviewed (approved):
The relnotes look nice and complete now :ok_hand:

tylerjereddy committed:
CI: PR 24207 revisions

* typo fix in linux_blas.yml

tylerjereddy commented:
I've updated with the latest backports and tried to pin `scipy-openblas*` in the one problem CI job (I don't think we need to be testing cutting edge releases there on the release branch anyway).

I did a little grepping and it seems that for wheel builds we have a hard pin on `scipy-openblas32==0.3.30.0.8` in `requirements/openblas.txt`, so I don't think there's a blocker there.

As for NumPy `2.4.0` -- I may just decide to move forward with RC2 since building wheels against it has no known issues (it is only the reverse of building against an older NumPy and using `2.4.0` at runtime). I'd say it would be suboptimal to do the final `1.17.0` release build against `2.4.0`, but the purpose of RC2 is to get feedback so I may prefer to avoid further delays if CI behaves here. I also thought about "banning" NumPy `2.4.0` in `pyproject.toml`, but then we'd basically just increase the chances of dealing with irrelevant-to-us (or at least, upstream) issues about the missing symbol when using those wheels against `2.4.0` at runtime.

tylerjereddy commented:
There were no outstanding review comments, the CI was green, and I saw nothing unusual in the diff, so merged.
